Userland/Libraries/LibThreading/BackgroundAction.h

```diff
@@ -34,51 +35,71 @@ class BackgroundActionBase {
 };
 
 template<typename Result>
-class BackgroundAction final : public Core::Object
-    , private BackgroundActionBase {
-    C_OBJECT(BackgroundAction);
+struct ResultWrapper : public Core::Object {
+    C_OBJECT(ResultWrapper);
 
-public:
-    void cancel()
+    ResultWrapper(Result& result)
+        : result(result)
     {
-        m_cancelled = true;
     }
 
-    bool is_cancelled() const
-    {
-        return m_cancelled;
-    }
+    Result& result;
+};
+
+// BackgroundAction has to inherit from Promise<Object> to allow it to be managed by EventLoop.
+// However, the internal result type is always ResultWrapper<Result>.
+template<typename Result>
+class BackgroundAction final : public Core::Promise<NonnullRefPtr<Core::Object>>
```

Comment:
This should not be a `public Promise`. It seems like any outside call to the public interface of `Promise` would be a programming error, e.g. setting `on_resolved` or calling `resolve()` would interfere with the inner workings of `BackgroundAction`, probably breaking it.

Consider making this a private descendant of Promise, and then making only some parts of the interface public, e.g. `using Promise<Result>::await();` or similar.

Reply:
It needs to be a public promise since EventLoop uses that interface. Otherwise, you can't put non-backgroundaction-promises into EventLoop (which seems silly; synchronous promises running some code in deferred_invoke seems like a useful feature)

Reply:
Okay, it needs to be a Promise, I agree.

But why public? The cast to `Promise` happens in the call to `EventLoop::add_job`, and that line is part of `BackgroundAction::BackgroundAction`, so it has access to a private base. What do I miss?

Reply:
My knowledge of C++ private inheritance semantics is only "don't use it except in very specific cases" so I just didn't know that that is / could be possible.

Reply:
Well it seems like a bad idea to keep it a public base, because the public API *looks* like it's meant to be used, especially the public world-writable `on_resolved` member, and doing so would break things. Here we can design the interface so that it's easier to get it right *and* do so in the intended way: private inheritance. So why not? :D